glib/src/unichar.rs

```diff
@@ -0,0 +1,238 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use std::mem::MaybeUninit;
+
+use crate::{
+    translate::{from_glib, IntoGlib},
+    UnicodeBreakType, UnicodeScript, UnicodeType,
+};
+
+mod sealed {
+    pub trait Sealed: Into<u32> {
+        fn from_unichar(unichar: u32) -> char;
+    }
+    impl Sealed for char {
+        #[inline]
+        #[track_caller]
+        fn from_unichar(unichar: u32) -> char {
+            // It's a glib bug if they give us an invalid unicode codepoint from one of their manipulation
+            // functions. It's safest if we just panic in that case
+            char::from_u32(unichar).expect("glib returned an invalid Unicode codepoint")
```

Comment:
FWIW I would go with `from_u32_unchecked()` here. If you really want the checks here then only enable them in `#[cfg(debug-assertions)]`.

The `expect()` here together with all the panic handling will considerably blow up the code size when working with these functions, and if glib gives us an invalid value then that's a bug in glib. We have to trust it to do the right thing in more critical places already.